Merge testharness.js

jugglinmike opened the pull request:
@jgraham I did this with two commits: one to remove the submodule, and one to re-introduce the code via `git subtree add`. [This seems to be the popular approach](https://stackoverflow.com/questions/28215244/convert-git-submodule-to-subtree), but it also produces an invalid state in the history where the `resources` directory is absent. I'm not sure if this is a problem; if so, let me know, and I'll try to find a more advanced solution.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/w3c/web-platform-tests/5536)
<!-- Reviewable:end -->

jgraham committed:
Merge pull request #53 from Ms2ger/reformat

Make whitespace and formatting more consistent in testharness.js.

zcorpan committed:
Make argument to step_func_done optional

Useful if you're expecting an event or callback but don't want to assert stuff.

zcorpan committed:
Make test(function(t) { t.add_cleanup(...) }) work

jgraham committed:
Merge pull request #58 from zcorpan/add-argument-to-test-function

Make test(function(t) { t.add_cleanup(...) }) work

foolip committed:
Remove media/sunflower.*

These are not used by web-platform-tests any longer:
https://github.com/w3c/web-platform-tests/pull/635

Ms2ger committed:
Merge pull request #60 from foolip/rm-sunflower

Remove media/sunflower.*

jgraham committed:
Don't try to halt tests when an exception progates to the top level

jgraham committed:
Prefer ERROR to TIMEOUT as a status

jgraham committed:
Drive-by cleanup of trailing whitespace.

jgraham committed:
Merge pull request #61 from w3c/jgraham/continue_on_exception

Don't try to halt tests when an exception progates to the top level

jgraham committed:
Pick up testharness settings from parent window for use in the runner

jgraham committed:
Merge pull request #63 from w3c/jgraham/testharness_properties

Pick up testharness settings from parent window for use in the runner

darobin committed:
when opening through a@target window.opener is defined but the SOP can trigger an exception if the origins don't match

darobin committed:
jshint pass

darobin committed:
space changes

Ms2ger committed:
Use localName rather than tagName.toLowerCase() when formatting Elements, to avoid unnecessary confusion.

darobin committed:
revert window.self, hadn't thought of the worker case - annoying though

jgraham committed:
Merge pull request #65 from w3c/jshint

jshint pass

jgraham committed:
Merge pull request #66 from Ms2ger/avoid-tagName

Use localName rather than tagName.toLowerCase() when formatting Elements...

jgraham committed:
Quick demo of single page tests.

jgraham committed:
fixup! Quick demo of single page tests.

darobin committed:
reformating to assuage James's sadistic tendencies

jgraham committed:
Merge pull request #64 from w3c/opener-exception

Exception on testharness_properties access

Manishearth committed:
Use https:// instead of git://

jgraham committed:
Merge pull request #68 from Manishearth/patch-1

Use https:// instead of git://

jgraham committed:
Update documentation for single page tests

jgraham committed:
Fix review issues

jgraham committed:
Add some more edge cases for single page tests.

jgraham committed:
fixup! Update documentation for single page tests

jgraham committed:
Add a force_timeout function to tests.

jgraham committed:
fixup! Add a force_timeout function to tests.

jgraham committed:
Merge pull request #70 from w3c/jgraham/force_timeout

Add a force_timeout function to tests.

jgraham committed:
Merge pull request #67 from w3c/jgraham/single_page

Add single-page test type.

jgraham committed:
Fixup from single page tests landing.

inexorabletash committed:
step_func propagates result, add unreached_func, Promise examples

In async tests, using testharness.js effectively requires that
callbacks are wrapped with step_func() calls so that assertion
failures are caught by the test harness. Promise chaining requires
that the callbacks return values, so step_func() needs to return the
value returned by the wrap function, in normal cases.

This adds some examples of Promise functionality (assuming ES6 promise
support in your browser) and the step_func fix, as well as
test.unreached_func which is handy for assigning to callbacks or
Promise catch clauses.

inexorabletash committed:
Remove copy/pasta

jgraham committed:
Merge pull request #71 from inexorabletash/inexorabletash/promise

step_func propagates result, add unreached_func, Promise examples

sideshowbarker committed:
Unbork IE support (JSON srlize/parse passed props)

sideshowbarker committed:
Added comment for IE workaround.

jgraham committed:
Merge pull request #72 from w3c/sideshowbarker/properties_pass

Unbork IE support (JSON srlize/parse passed props)

zcorpan committed:
Include a copy-pastable template for new tests

jgraham committed:
Merge pull request #73 from w3c/copy-pastable-template-in-comment

Include a copy-pastable template for new tests

jgraham committed:
Move docs to their own files and format as markdown.

jgraham committed:
Merge pull request #74 from w3c/jgraham/move_docs

Move docs to their own files and format as markdown.

tobie committed:
Update WebIDL submodule to latest release.

jgraham committed:
Merge pull request #76 from tobie/update-webidl2

Update WebIDL submodule to latest release.

jgraham committed:
Set harness status correctly when tests time out.

jgraham committed:
fixup! Set harness status correctly when tests time out.

jgraham committed:
Merge pull request #77 from w3c/timeout_status

Set harness status correctly when tests time out.

jgraham committed:
Make harness status more obvious in default output.

jgraham committed:
Merge pull request #78 from w3c/jgraham/harness_status

Make harness status more obvious in default output.

jgraham committed:
Don't set the output log element until we are done loading.

jgraham committed:
Merge pull request #79 from w3c/jgraham/log_wait_load

Don't set the output log element until we are done loading.

asankah committed:
Support testing in a worker.

This change:
* Allows testharness.js to be imported into a worker. Currently
  dedicated workers, shared workers and service workers are supported.
* Allows test results from a worker to be imported to a client document.
* Introduces promise_test() that handles some boilerplate for testing
  with Promises.

Within a worker, the test harness can be imported via:

  importScripts('/resources/testharness.js');

Once imported, the worker can invoke tests using test(), async_test(),
etc. The testharness sends tests results from the worker to any
connected client document that called fetch_tests_from_worker(). See
apisample14.html, apisample15.html and apisample16.html.

Ms2ger committed:
Split up IdlInterface.prototype.test_members.

Ms2ger committed:
Add tests for stringifiers to idlharness.

Ms2ger committed:
fixup! Split up IdlInterface.prototype.test_members.

plehegar committed:
Merge pull request #83 from Ms2ger/idlharness-stringifier

Add tests for stringifiers to idlharness.

zcorpan committed:
Point to api.md where the documentation is

Ms2ger committed:
Update the configurability of interface objects' length.

Ms2ger committed:
Check for LenientThis when testing attribute getters.

Ms2ger committed:
Test for calling attribute setters on an incorrect object.

jgraham committed:
Merge pull request #87 from Ms2ger/configurable-interface-object-length

Update the configurability of interface objects' length.

zcorpan committed:
Merge pull request #89 from Ms2ger/lenientthis

Support [LenientThis] extended attribute (fixes #88).

Ms2ger committed:
Add support for ByteString and USVString to IdlArray#assert_type_is.

sideshowbarker committed:
Point readme to latest docs location.

zqzhang committed:
Merge pull request #91 from w3c/sideshowbarker/docs-fix

Point readme to latest docs location.

Ms2ger committed:
Add a pointer to an issue.

Ms2ger committed:
Add support for ByteString and USVString to create_suitable_object.

zcorpan committed:
Merge pull request #90 from Ms2ger/strings

Add support for ByteString and USVString to IdlArray#assert_type_is.

jgraham committed:
Merge pull request #82 from asankah/support-testing-in-a-service-worker

Support testing in a service worker

asankah committed:
Add documentation for promise_test() and worker tests.

zcorpan committed:
Actually substitute expected and actual

zcorpan committed:
Merge pull request #94 from w3c/zcorpan/assert_array_equals

Actually substitute expected and actual

asankah committed:
Address review comments

jgraham committed:
Merge pull request #93 from asankah/update-docs-for-workers

Add documentation for promise_test() and worker tests.

luser committed:
Add support for DOMTimeStamp and DOMHighResTimeStamp to idlharness.js

zqzhang committed:
Merge pull request #98 from luser/idlharness-timestamps

Add support for DOMTimeStamp and DOMHighResTimeStamp to idlharness.js

Ms2ger committed:
Ensure Test.set_status is not called with a non-string message argument.

This would cause an error in escape_html later, when it tries to call
.replace() on a non-string value.

Ms2ger committed:
Don't assume the exception is an object in assert_throws.

Ms2ger committed:
Implement a promise_rejects function to match assert_throws.

Ms2ger committed:
Update assert_throws to support more errors defined in WebIDL.

sideshowbarker committed:
Merge pull request #100 from w3c/more-errors

Update assert_throws to support more errors defined in WebIDL.

sideshowbarker committed:
Merge pull request #99 from Ms2ger/rejection

Implement a promise_rejects function to match assert_throws.

Ms2ger committed:
Update idlharness to work in workers.

Ms2ger committed:
Merge pull request #101 from Ms2ger/idlharness-worker

Update idlharness to work in workers; r=jgraham

Ms2ger committed:
Handle members of global interfaces.

Ms2ger committed:
Hook up interface object proto chains as if they were ES6 classes.

This covers <https://www.w3.org/Bugs/Public/show_bug.cgi?id=23133>.

Ms2ger committed:
Use IdlInterface#is_callback throughout.

The Callback extended attribute is long gone.

Ms2ger committed:
Test callback interfaces.

Ms2ger committed:
Update test for the interface prototype object.

Ms2ger committed:
Disable some tests for callback interfaces that don't have an interface object.

Ms2ger committed:
Remove IDLException.

WebIDL no longer has support for arbitrary exception types.

Ms2ger committed:
Merge pull request #102 from Ms2ger/23133-chains

Hook up interface object proto chains as if they were ES6 classes.

Ms2ger committed:
Merge pull request #104 from Ms2ger/globals

Handle members of global interfaces; r=zcorpan

jwatt committed:
Add assert_between_exclusive and assert_between_inclusive methods to testharness.js

zcorpan committed:
Merge pull request #103 from Ms2ger/callback-interfaces

Support callback interfaces.

Ms2ger committed:
Collapse IdlExceptionOrInterface and IdlInterface.

jwatt committed:
Add the new assert_between_exclusive and assert_between_inclusive functions to the API documentation

jgraham committed:
Merge pull request #107 from jwatt/master

Add assert_between_exclusive and assert_between_inclusive methods to testharness.js

jwatt committed:
Fix the assertion text in assert_less_than_equal so that assertion failures are attributed to this function

Ms2ger committed:
Merge pull request #109 from jwatt/master

Fix the assertion text in assert_less_than_equal so that assertion failures are attributed to this function; r=Ms2ger

zcorpan committed:
Merge pull request #105 from Ms2ger/rm-exn

Remove support for IDL exceptions.

jgraham committed:
Improve handling of stack traces in results collection.

Ensure that failing asserts give a stack.
Store the stack as a seperate property of the result objects (this
is a protocol change).
Nicer formatting of the stack in the default output

jgraham committed:
Merge pull request #110 from w3c/jgraham/stack_trace

Improve handling of stack traces in results collection.

jwatt committed:
Add EventWatcher helper to testharness.js

It is very difficult to test that a series of events occurs in a very specific
order, while also checking that unexpected events do not occur between any of
the expected events. This helper can create Promises that resolve when the
expected events occur, and reject when unexpected events occur, which makes
the process of testing such events much more feasible.

jgraham committed:
Merge pull request #111 from jwatt/master

Add EventWatcher helper to testharness.js

jwatt committed:
Expose the EventWatcher constructor function

jgraham committed:
Merge pull request #113 from jwatt/master

Expose EventTester constructor function

jgraham committed:
Don't error out trying to access cross-origin functions when document.domain got set.

Setting document.domain changes the origin of the current document to be different
to the origin it had before. This invalidates the cache of origins, so in this case
we need to catch the error and make sure we don't try to call a cross-origin function.

g-ortuno committed:
Change promise_test to run sequentially.

The goal is to make tests that share state (across the tests, or in the object under test) be deterministic.

Ms2ger committed:
Merge pull request #114 from w3c/jgraham/document_domain_dispatch

Don't error out trying to access cross-origin functions when document.domain got set; r=Ms2ger

bzbarsky committed:
Improve the handling of unforgeable interfaces and Global/PrimaryGlobal interfaces, for which operations don't live where they do for normal interfaces

Ms2ger committed:
Merge pull request #116 from bzbarsky/fix-unforgeables-and-globals

Improve the handling of unforgeable interfaces and Global/PrimaryGlobal interfaces, for which operations don't live where they do for normal interfaces; r=Ms2ger

bzbarsky committed:
Make the unforgeable setup in idlharness a bit simpler to work with

Ms2ger committed:
Merge pull request #117 from bzbarsky/better-unforgeable

Make the unforgeable setup in idlharness a bit simpler to work with; r=Ms2ger

jgraham committed:
Fix typo getting error stack in worker.

zqzhang committed:
Merge pull request #120 from w3c/jgraham/worker_error_typo

Fix typo getting error stack in worker.

jgraham committed:
Add something that dumps the test result JSON to a hidden script element when the run is complete.

This is experimental until we have a better way to integrate with saucelabs

inexorabletash committed:
Replace ES6 arrow expressions with boring old functions

jgraham committed:
Merge pull request #121 from inexorabletash/inexorabletash/noarrows

Replace ES6 arrow expressions with boring old functions

kojiishi committed:
Fix AssertionError.get_stack for IE

wanderview committed:
Fall back to ServiceWorker.postMessage() when MessageChannel not available. (gecko bug 1158347)

wanderview committed:
Tweak comments and indentation per review feedback.

jgraham committed:
Merge pull request #123 from wanderview/bug1158347_messagechannel

Fall back to ServiceWorker.postMessage() when MessageChannel not availab...

Ms2ger committed:
Merge pull request #122 from kojiishi/get_stack-for-IE

Fix AssertionError.get_stack for IE

jgraham committed:
Add [test.]step_timeout function.

This runs a test step after a given timeout, which is scaled by the
timeout multiplier.

Ms2ger committed:
Correct the expected length of overloaded operations.

Fixes #86.
Fixes #95.
Fixes #124.

Ms2ger committed:
Correct the test that calls methods with too few arguments.

Fixes #118.
Fixes #125.

Ms2ger committed:
Merge pull request #128 from Ms2ger/length

Fix bugs related to optional and variadic arguments; r=nox

jgraham committed:
Set correct this value in worker structured clone function.

jgraham committed:
Add ability to subscribe to only certain events via postMessage.

Ms2ger committed:
Merge pull request #131 from w3c/jgraham/worker_struct_clone

Set correct this value in worker structured clone function; r=Ms2ger

jgraham committed:
Merge pull request #132 from w3c/jgraham/message_events

Add ability to subscribe to only certain events via postMessage.

DLehenbauer committed:
Fix potential infinite loop in AssertionError.get_stack().

If testharness.js did not reside under /resources/, the loop to find
the first frame originating from the harness would never terminate.

Fixes #129.

Ms2ger committed:
Merge pull request #130 from MicrosoftEdge/fix_infinite_loop

Fix potential infinite loop in AssertionError.get_stack(); r=Ms2ger

plehegar committed:
Added BSD license

jgraham committed:
Merge pull request #133 from w3c/license

Added BSD license

jgraham committed:
Merge pull request #115 from g-ortuno/sequential-promise-tests

Turn promise tests into sequential promise tests.

inexorabletash committed:
Fix handling of assertions in workers

inexorabletash committed:
Move apisamples to examples directory.

Ms2ger committed:
Merge pull request #138 from inexorabletash/inexorabletash/sample-dir

Move apisamples to examples directory; r=Ms2ger

inexorabletash committed:
braces

jgraham committed:
Merge pull request #136 from inexorabletash/inexorabletash/worker-script-url

Fix handling of assertions in workers

jgraham committed:
Merge pull request #127 from w3c/jgraham/after_timeout

Add [test.]step_timeout function.

bboyle committed:
syntax highlighting for js and html code blocks

- use fenced code blocks to enable GFM syntax highlighting: https://help.github.com/articles/github-flavored-markdown/#syntax-highlighting
- remove whitespace indent to reduce horizontal scrolling in code blocks

Ms2ger committed:
Merge pull request #145 from bboyle/bboyle/api-doc-syntax

syntax highlighting for js and html code blocks; r=Ms2ger

cvrebert committed:
docs/api.md: Fix link to "List of Assertions" section

cvrebert committed:
testharnessreport.js: Avoid having a literal </script> within the source code

Ms2ger committed:
Merge pull request #146 from cvrebert/patch-1

docs/api.md: Fix link to "List of Assertions" section; r=Ms2ger

jgraham committed:
Merge pull request #147 from cvrebert/patch-2

testharnessreport.js: Avoid having a literal </script> within the code

Ms2ger committed:
Simplify legacycaller detection.

WebIDLParser now supports it, with a boolean legacycaller member.

Ms2ger committed:
Allow using assert_inherits with function objects.

idlharness hits this case for objects with legacycallers.

jgraham committed:
Merge pull request #149 from Ms2ger/assert_inherits-function

Allow using assert_inherits with function objects.

plehegar committed:
Sync'ed up with upstream develop branch of webidl2.js

Ms2ger committed:
Merge pull request #150 from w3c/plh/webidl2-sync

Sync'ed up with upstream develop branch of webidl2.js; r=Ms2ger

plehegar committed:
Merge pull request #148 from Ms2ger/legacycaller

Simplify legacycaller detection.

gsnedders committed:
Fix #59: stop cross-origin objects from throwing in harness when creating failed assertion messages.

gsnedders committed:
Distinguish the various float types.

gsnedders committed:
Fix #92: distinguish text types.

Ms2ger committed:
Reinstate InUseAttributeError.

It came into use again for setAttributeNode and friends.

Ms2ger committed:
Merge pull request #156 from Ms2ger/InUseAttributeError

Reinstate InUseAttributeError; r=Manishearth

dontcallmedom committed:
Adapt throw test for Promise-returning operations

annevk committed:
Variable typo in `EventWachter` example

dontcallmedom committed:
Fix whitespace indent

dontcallmedom committed:
Use promise_rejects to test rejection

nox committed:
Test the interface object name

Ms2ger committed:
Merge pull request #166 from nox/interface-object-name

Test the interface object name; r=Ms2ger

edmilsonefs committed:
Fix typos in Testharness.js API Documentation

jgraham committed:
Update to latest webidl2

jgraham committed:
Merge pull request #169 from edmilsonefs/edmilsonefs/typos

Fix typos in Testharness.js API Documentation

jgraham committed:
Update to develop branch of webidl2

Ms2ger committed:
Remove some references to IDL exceptions.

Ms2ger committed:
Don't try to handle the case where test_member_const is called with no constants.

Ms2ger committed:
Always test the type of the getter for attributes on global objects.

Ms2ger committed:
Update the comments in test_member_operation.

Ms2ger committed:
Merge pull request #172 from w3c/idlharness

Various idlharness improvements; r=nox

DLehenbauer committed:
Issue #157: Fix HierarchyRequestError with XHTML and SVG documents

DLehenbauer committed:
Avoid unintentional use of global var 'data'

DLehenbauer committed:
Fix /*global*/ decls and nearby whitespace

(VS Code did not understand the comma delimited list.)

Ms2ger committed:
Merge pull request #160 from MicrosoftEdge/fix-xhtml-svg-dom-hierarchy-error

Issue #157: DOM hierarchy error with XHTML and SVG; r=Ms2ger

dontcallmedom committed:
Use asynchronous test without creating a new test object

per https://github.com/w3c/testharness.js/pull/161#issuecomment-159901984 and https://github.com/zqzhang

dontcallmedom committed:
Fix indents

Per @zqzhang feedback

Commit:
Merge pull request #161 from w3c/promise-throw

Adapt throw test for Promise-returning operations

zcorpan committed:
Merge pull request #154 from gsnedders/cross_origin_assert_safety

Fix #59: stop cross-origin objects from throwing in harness when creating failed assertion messages

Ms2ger committed:
Fix typo in the readme.

jgraham committed:
Merge pull request #181 from w3c/readme

Fix typo in the readme.

scheib committed:
Allow longer strings when emitting errors, truncate at 1000 vs 60.

format_value truncates strings for readability, but was too aggressive. More recent APIs implementation emit useful error messages, e.g. Web Bluetooth may emit a message such as:

SyntaxError: Failed to execute 'getCharacteristic' on 'BluetoothGATTService': Invalid Characteristic name: 'heart_rate'. It must be a valid UUID alias (e.g. 0x1234), UUID (lowercase hex characters e.g. '00001234-0000-1000-8000-00805f9b34fb'), or recognized standard name from https://developer.bluetooth.org/gatt/characteristics/Pages/CharacteristicsHome.aspx e.g. 'aerobic_heart_rate_lower_limit'.

At least 110 characters are necessary to understand the reason a test is failing. The truncation shouldn't make tests hard to understand or diagnose. A well formed and performing test will not emit long strings, so this is primarily for diagnosing when something isn't working.

Removing truncation may result in very large objects still completely spamming the output. Thus, increasing to 1000, as an order of magnitude larger than what is known to be useful seems an appropriate solution.

AlexeyKor committed:
Issue #179: Fix step_timeout

jdm committed:
Merge pull request #185 from MicrosoftEdge/fix-step-timeout

Issue #179: Fix step_timeout

jdm committed:
Merge pull request #155 from gsnedders/distinguish_types

Distinguish float/string types

Ms2ger committed:
Replace some more characters in format_value.

jgraham committed:
Merge pull request #188 from w3c/format_value

Replace some more characters in format_value.

mkruisselbrink committed:
Add description parameter to promise_rejects. (#191); r=Ms2ger

All assert methods take a description parameter, so for consistency also
add one to promise_rejects.

This fixes #190

jgraham committed:
Merge pull request #184 from scheib/increase-truncate-limit

Allow longer strings when emitting errors, truncate at 1000 vs 60.

DigiTec committed:
This change adds some basic examples of generate_tests. While working
on many DOM tests there are often times when you want an exhaustive
round of tests, such as for parameter range validation. I found that
we often break these apart and repeat the calls over and over again
when a simple generate_tests would simplify the code greatly.

annevk committed:
Escape the URL passed to RegEx (#193)

jgraham committed:
Merge pull request #194 from DigiTec/master

Add some basic examples of generate_tests.

annevk committed:
Encourage good test titles (#195)

zcorpan committed:
Use more real-world examples (#196)

* Use more real-world examples

Just assert_true(true) is not so helpful in understanding how one is supposed to use this API.

Also document step_func_done.

* Change timeout back to 2000

* Say something about timeouts needing to be long

zcorpan committed:
Typo fix in docs for assert_approx_equals

NavidZ committed:
Add the null check of get_script_url back (#197)

* Add the null check of get_script_url back

The last change to this line removed the null check
of get_script_url. Adding it back and keeping the
regex functionality that was added in the previous change.

* Splitting the line

cpsgchue committed:
add checking of WorkerGlobalScope for DedicatedWorkerGlobalScope (#174); r=Ms2ger

youennf committed:
Adding iterable testing

Ms2ger committed:
Consider [Exposed] in idlharness. (#202)

TakayoshiKochi committed:
Remove trailing whitespace.

Commit:
Merge pull request #203 from TakayoshiKochi/whitespace

Remove trailing whitespace.

tidoust committed:
Fix typo in promise_rejects example

The promise_rejects function expects a Promise as third argument, not a
function. The "bar" function needs to be called.

jdm committed:
Merge pull request #206 from tidoust/promise_rejects

Fix typo in promise_rejects example

zcorpan committed:
Run stringifier test in idlharness for attributes (#205); r=Ms2ger

Ref. https://github.com/w3c/web-platform-tests/pull/3545

mkruisselbrink committed:
Don't call test.step for promise_tests too soon. (#209)

Calling test.step starts the timeout for the test, for promise tests this
shouldn't happen until the the previous promise_test finished, and the
current test is actually being started.

This fixes #208

mantaroh committed:
Fix EventWatcher example parameter, we should specify the test object in constructor.

dontcallmedom committed:
Do no choke on Promise attributes

close #34

cvrebert committed:
Make get_stack() cope with `Error.stack` not being supported

For increased compatibility, particularly with old browser versions.

tidoust committed:
Add "NotAllowedError" to the list of DOMException names

"NotAllowedError" is defined in WebIDL Editor's Draft:
https://heycam.github.io/webidl/#notallowederror

The error is used in the Presentation API specification in particular:
https://w3c.github.io/presentation-api/#terminology

Ms2ger committed:
Try to catch misuse of promise_test earlier. (#220)

ayg committed:
Callback interface object prototype is now Function.prototype

Update for heycam/webidl#156.

gsnedders committed:
Set a test to ERROR if it has an unhandled promise rejections (#219)

See <http://www.w3.org/mid/556aa889-0ad1-9789-f1db-a148fe7a9929@nottheoilrig.com>.

ayg committed:
Add fold markers to new idlharness functions

In vim, these make the file much easier to navigate: function bodies are
collapsed until you expand them (and zc recollapses them).  I don't know
if any other editors recognize them.  Since the whole file uses them, we
may as well be consistent.

cvrebert committed:
Specify useCapture argument in addEventListener calls (#215)

pwnall committed:
Unblock test promise chain when done() is called early.

This makes it possible to end a promise_test by calling Test.done().
Previously, that would work for a single test, but would get any
follow-up test stuck waiting for the global promise chain to resolve.

zhaobin2016 committed:
check if promise object has then property with in operator

jugglinmike committed:
Correct typos in documentation (#230)

bzbarsky committed:
Partially update idlharness to Promise-returning attributes now rejecting instead of throwing on invalid this and the like.

sfuqua committed:
Addition of fetch_tests_from_window.

The new function is based on existing fetch_tests_from_worker.
This allows a primary test window to combine tests from multiple child
windows (e.g., opened windows or iframes) into one test suite. The
existing RemoteWorker object was restructured into RemoteContext, which
abstracts the usage of RemoteTest so that the same code can be used for
both workers and alternate windows.

Commit:
idlharness: Swap result and expectation in one of the iterable tests

The pair iterable test was passing its expected value as the actual value,
leading to some confusing error messages when an failure occurred.

jdm committed:
Merge pull request #235 from rakuco/idlharness-iterable-expectation

idlharness: Swap result and expectation in one of the iterable tests

sfuqua committed:
Add doc and sample for fetch_tests_from_window.

sfuqua committed:
Add additional fetch_tests_from_window sample.

sfuqua committed:
Minor documentation update.

sfuqua committed:
Update documentation.

JosephPecoraro committed:
Fix API documentation (#237)

 - invalid code examples
 - clean up trailing punctuation

Ms2ger committed:
Update to LegacyArrayClass.

sfuqua committed:
Update WindowTestEnvironment with 'getmessages'.

This allows other interested windows to query the status of
tests that may have already finished by queueing outbound
messages in a similar fashion to WorkerTestEnvironment.

bzbarsky committed:
Fix interface [Exposed] handling to handle string args, not just list args, and implement [Exposed] handling for interface members.

bzbarsky committed:
Add proper handling of FrozenArray types and fix the handling of sequence types to pass in the right "inner" type.

bzbarsky committed:
Fix iterable testing to add actual members to our list of members; otherwise all sorts of stuff breaks (#243)

jgraham committed:
Ensure that we stop the run if we get an uncaught error. (#152)

Without this change it was typically necessary for the harness
to timeout before the result would be reported. In some cases
e.g. with explicit_done set, this would never happen, leading
to the appearance of a hang.

sfuqua committed:
Change 'const' to 'var'.

jugglinmike committed:
Fail in the presence of duplicate test names (#240)

Test names are presumed to be unique within test files--this allows
consumers to use them for identification purposes. Duplicated names
violate this expectation and should therefore be rejected. Previously,
the default "reporter" implementation identified this case only
informally (through modification to the test document)--the error was
not communicated to whatever "runner" process initiated test execution.

Formally report an error in the presence of duplicate test names. Remove
the equivalent code in the default "reporter" implementation (as the
error is now communicated via the typical "test failure" mechanism).

sfuqua committed:
Add null check

annevk committed:
Remove usage of ancestorOrigins

WebKit throws the same exception as other browsers these days and
ancestorOrigins might become less reliable due to
https://github.com/whatwg/html/issues/1918.

ayg committed:
get_harness_url() returns undefined on failure, not null

This resulted in trying to fetch "undefinedtestharness.css", which is
probably a 404 on most servers.  Strict equality checking is not always
an advantage.

ayg committed:
Support unions in idlharness

mikewest committed:
Add 'CancelationError'.

Once https://github.com/heycam/webidl/pull/339 lands, we'll need this.

Thanks to @annevk for telling me that this list existed. :)

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/221146451)

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/221146452)

jgraham commented:
Hmm, you might need to check that the webidl2 submodule is correctly set up (or maybe convert that to a subtree also, but without the history and with the expectation that we update it sometimes).

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/221463738)

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/221463739)

jugglinmike committed:
Manually reconfigure git submodule

The previous subtree introduction included a submodule, but git does not
currently transfer submodules correctly in this case. Manually update
the submodule configuration to reflect the relationship between these
directories.

w3c-bots commented:
These tests are now available on [w3c-test.org](https://w3c-test.org/submissions/5536/)

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/221465492)

# Firefox (nightly channel) #
Testing web-platform-tests at revision 59620d31ff65659c7e8c51bf09e0f725629e9992
Using browser at version BuildID 20170412100254; SourceStamp f40e24f40b4c4556944c762d4764eace261297f5
Starting 10 test iterations

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/221465493)

# Chrome (unstable channel) #
Testing web-platform-tests at revision 59620d31ff65659c7e8c51bf09e0f725629e9992
Using browser at version 59.0.3067.0 dev
Starting 10 test iterations

jugglinmike committed:
[check stability] Allow changes to be ignored

Because this extension introduces a operation that can be described as
"skipping", the internal variable `skip_dirs` is somewhat ambiguous.
Rename to `skip_tests` to help differentiate from the new behavior
(implented as `ignore_changes`).

jugglinmike commented:
Alright @jgraham, this is ready for another review.

jgraham commented on check_stability.py:
```diff
@@ -17,6 +18,8 @@
 
 import requests
 
+_CONFIG_FILE = "check_stability.ini"
```
Instead of making this a global, I might be tempted to make it a command line option with a default value.

jgraham commented on check_stability.py:
```diff
@@ -782,7 +804,13 @@ def main():
+        if files_ignored:
+            logger.info("Ignoring %s changed files:\n%s" % (len(files_ignored),
+                                                            "".join(" * %s\n" % item for item in files_ignored)))
+        else:
```
I don't think this else clause is generally helpful

jugglinmike commented:
Thanks for the feedback, @jgraham! I've made the changes you've requested.

sfuqua committed:
Start MessagePort for SharedWorkers

jugglinmike committed:
Remove `resources` submodule

This code will be re-introduced in a subsequent commit using a git
subtree.

jugglinmike committed:
Add 'resources/' from commit 'acae62d14f75fc9c0de6ffb6bba37a8cb8ed8112'

git-subtree-dir: resources
git-subtree-mainline: 9c1de7a747fde4bbb467caf2b974dfd21c30ac2f
git-subtree-split: acae62d14f75fc9c0de6ffb6bba37a8cb8ed8112

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/222954124)

# Lint #

Passed

jugglinmike commented:
@jgraham I've updated this patch to include the subtree as of https://github.com/w3c/web-platform-tests/pull/5591

jgraham commented:
Merged this in 39d07eb

jugglinmike commented:
Hooray! Thanks, James!
